examples/esp32/xively_demo/main/xively_if.c

```diff
@@ -0,0 +1,617 @@
+{
+    int retval = 0;
+
+    assert( NULL != topic_name );
+    assert( NULL != dst );
+
+    retval = snprintf( dst, dst_len, "xi/blue/v1/%.64s/d/%.64s/%.64s",
+                       xif_mqtt_device_info.xi_account_id,
+                       xif_mqtt_device_info.xi_device_id, topic_name );
+
+    if ( ( retval >= dst_len ) || ( retval < 0 ) )
+    {
+        return -1;
+    }
+    return 0;
+}
+
+int xif_init( void )
+{
+    xi_state_t xi_ret_state = XI_STATE_OK;
+
+    xif_requests_event_group_handle = xEventGroupCreate();
+    if( NULL == xif_requests_event_group_handle )
+    {
+        printf( "\n[XIF] Failed to allocate FreeRTOS heap for the requests event group" );
+        return -1;
+    }
+
+    xi_ret_state = xi_initialize( xif_mqtt_device_info.xi_account_id,
+                                  xif_mqtt_device_info.xi_device_id );
+    if ( XI_STATE_OK != xi_ret_state )
+    {
+        printf( "\n[XIF] Failed to initialize MQTT library" );
+        return -1;
+    }
+
+    xif_context_handle = xi_create_context();
+    if ( 0 > xif_context_handle )
+    {
+        printf( "\n[XIF] Failed to create a valid libxively context. Returned: %d",
+                xif_context_handle );
+        return -1;
+    }
+    return 0;
+}
+
+/* @retval: -1: xi_connect failed
+ *           0: 0 xi_connect OK
+ *           1: xi_connect failed because the connection was already initialized
+ */
+int xif_connect( void )
+{
+    xi_state_t ret_state = XI_STATE_OK;
+    printf( "\n[XIF] Connecting to MQTT broker:" );
+    printf( "\n[XIF] \tAccount ID: %s", xif_mqtt_device_info.xi_account_id );
+    printf( "\n[XIF] \tDevice  ID: %s", xif_mqtt_device_info.xi_device_id );
+    printf( "\n[XIF] \tAccount Password: (REDACTED)" );
+    ret_state = xi_connect( xif_context_handle, xif_mqtt_device_info.xi_device_id,
+                            xif_mqtt_device_info.xi_device_pwd, 10, 0, XI_SESSION_CLEAN,
+                            &xif_on_connected );
+
+    if( XI_ALREADY_INITIALIZED == ret_state )
+    {
+        return 1;
+    }
+    else if ( XI_STATE_OK != ret_state )
+    {
+        printf( "\n[XIF] Error in xi_connect! Retcode: %d", ret_state );
+        return -1;
+    }
+    return 0;
+}
+
+int xif_subscribe( void )
+{
+    xi_state_t ret_state = XI_STATE_OK;
+    printf( "\n[XIF] Subscribing to MQTT topics" );
+    ret_state            = xi_subscribe( xif_context_handle, xif_mqtt_topics.led_topic,
+                              XI_MQTT_QOS_AT_MOST_ONCE, &xif_led_topic_callback, NULL );
+    if ( XI_STATE_OK != ret_state )
+    {
+        return -1;
+    }
+    return 0;
+}
+
+void xif_publish_button_state( int button_state )
+{
+    char msg_payload[2] = "";
+    if( !xif_is_connected() )
+    {
+        return;
+    }
+
+    /* GPIO0 is pulled-up, so the input level is 0 when pressed. If your backend
+    expects 1 to mean "button pressed", you can negate the button input here */
+    ( button_state == 1 ) ? ( strcpy( msg_payload, "1" ) )
```

Comment:
So for product launcher visualization I found it's more helpful to publish a toggle value.  1 when the button is pressed, 0 when it's pressed again, 1 the next time, etc.

Reply:
100% agreed. That's the current behaviour in this demo too. There's a `virtual_switch` variable in `main.c::app_gpio_interrupts_handler_task()` that keeps track of the state.

I guess I have to find a better name for this function.